buildpacks/nodejs-corepack/src/main.rs

```diff
@@ -0,0 +1,118 @@
+#![warn(unused_crate_dependencies)]
+#![warn(clippy::pedantic)]
+#![warn(clippy::cargo)]
+#![allow(clippy::module_name_repetitions)]
+
+use heroku_nodejs_utils::package_json::{PackageJson, PackageJsonError};
+use layers::{ManagerLayer, ShimLayer};
+use libcnb::build::{BuildContext, BuildResult, BuildResultBuilder};
+use libcnb::data::build_plan::BuildPlanBuilder;
+use libcnb::data::layer_name;
+use libcnb::detect::{DetectContext, DetectResult, DetectResultBuilder};
+use libcnb::generic::GenericMetadata;
+use libcnb::generic::GenericPlatform;
+use libcnb::layer_env::Scope;
+use libcnb::{buildpack_main, Buildpack, Env};
+use libherokubuildpack::log::log_header;
+
+#[cfg(test)]
+use libcnb_test as _;
+#[cfg(test)]
+use test_support as _;
+#[cfg(test)]
+use ureq as _;
+
+mod cfg;
+mod cmd;
+mod errors;
+mod layers;
+
+pub(crate) struct CorepackBuildpack;
+
+impl Buildpack for CorepackBuildpack {
+    type Platform = GenericPlatform;
+    type Metadata = GenericMetadata;
+    type Error = CorepackBuildpackError;
+
+    fn detect(&self, context: DetectContext<Self>) -> libcnb::Result<DetectResult, Self::Error> {
+        // Corepack requires the `packageManager` key from `package.json`.
+        // This buildpack won't be detected without it.
+        let pkg_json_path = context.app_dir.join("package.json");
+        if pkg_json_path.exists() {
+            let pkg_json =
+                PackageJson::read(pkg_json_path).map_err(CorepackBuildpackError::PackageJson)?;
+            cfg::get_supported_package_manager(&pkg_json).map_or_else(
+                || DetectResultBuilder::fail().build(),
+                |pkg_mgr| {
+                    DetectResultBuilder::pass()
+                        .build_plan(
+                            BuildPlanBuilder::new()
+                                .requires("node")
+                                .requires(&pkg_mgr)
+                                .provides(pkg_mgr)
+                                .build(),
+                        )
+                        .build()
+                },
+            )
+        } else {
+            DetectResultBuilder::fail().build()
+        }
+    }
+
+    fn build(&self, context: BuildContext<Self>) -> libcnb::Result<BuildResult, Self::Error> {
+        let pkg_mgr = PackageJson::read(context.app_dir.join("package.json"))
+            .map_err(CorepackBuildpackError::PackageJson)?
+            .package_manager
+            .ok_or(CorepackBuildpackError::PackageManager)?;
```

Comment:
Nit and nothing that blocks this PR: I find `PackageManager` as an error value weird - it doesn't include what's the actual issue in the name.

Reply:
Fixed in [603921e](https://github.com/heroku/buildpacks-nodejs/pull/418/commits/603921ed2d4d8c71522eec354374133aebae2d13).